js/ui/messageTray.js

```diff
@@ -552,16 +552,16 @@ class NotificationBanner extends Calendar.NotificationMessage {
 
         this._buttonBox.add(button);
         button.connect('clicked', () => {
-            if (!this.notification.resident) {
+            callback();
+
+            if (this.notification && !this.notification.resident) {
```

Comment:
This seems plausible but what I don't really understand is how `this.notification` can become `null`/`undefined` – I don't see any assignments to it.

Reply:
I am building up on @andrunko 's investigation mentioned in the commit message: https://github.com/endlessm/gnome-shell/commit/9ba4a2b106a28d93a2bef635fe453e79ae9d253d

Reply:
I wonder how this interacts with the 'done-displaying' signal emitted within this block – will this have been emitted by something higher up the call stack? Does it matter that it's not emitted?

Reply:
Quick look suggests the notification should hide(actually terminate-displaying...) if the user  invokes its action by clicking of a button. That's why the `done-displaying` signal.

OTOH, if the notification has `setResident` set, the notification is *not* finished-displaying yet (because further information might be in-flight and notification is updated realtime), in that case, we don't want to emit `done-displaying` signal and destroy the notification.

Reply:
As @wjt pointed out we should make sure "done-displaying" is emitted even if the notification got destroyed in the callback (which is the case).

Before invoking the callback, save the value of "this.notification.resident" and after invoking the callback emit "done-displaying" regardless of whether the notification got destroyed in the callback?

Please also make sure the undo action still works with this change when removing icons from the desktop.

Reply:
@uajain missed your last reply but I still think we should emit "done-displaying" if the notification is not resident (even if it got destroyed in the callback), please verify and also check about the undo action...

Reply:
Much deeper issue than I thought. But also reproduce-able on upstream version. 
https://gitlab.gnome.org/GNOME/gnome-shell/issues/2140

I think we should keep the revert (respecting [this](https://phabricator.endlessm.com/T13041#796575)) and fix the issue for "Undo" use-case upstream itself.